x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -142,10 +142,13 @@ public RBACEngine(
     @Override
     public void resolveAuthorizationInfo(RequestInfo requestInfo, ActionListener<AuthorizationInfo> listener) {
         final Authentication authentication = requestInfo.getAuthentication();
-        rolesStore.getRoles(
-            authentication,
-            listener.delegateFailureAndWrap((l, roleTuple) -> l.onResponse(new RBACAuthorizationInfo(roleTuple.v1(), roleTuple.v2())))
-        );
+        rolesStore.getRoles(authentication, listener.delegateFailureAndWrap((l, roleTuple) -> {
+            if (roleTuple.v1() == Role.EMPTY_RESTRICTED_BY_WORKFLOW || roleTuple.v2() == Role.EMPTY_RESTRICTED_BY_WORKFLOW) {
+                l.onResponse(DeniedAuthorizationInfo.INSTANCE);
```

Comment:
I don't think this is an issue, but would like to write down for prosperity: `v1` is the run-as user's role, `v2` is the authenticating user's role. If `v1` is empty and restricted while `v2` is a proper role, one might argue that `v2` should be able to run-as `v1` (i.e. pass `authorizeRunAs`) then fail. 

I think it can be argued both ways because:

We have a similar situation where request fails before authorization happens if [run-as user is disabled](https://github.com/elastic/elasticsearch/blob/31a47867428f1f53bfa72b049fd5e082a4fef234/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L320-L321). 

At the same time, we also have request fails at authorization if the [run-as user does not exist](https://github.com/elastic/elasticsearch/blob/31a47867428f1f53bfa72b049fd5e082a4fef234/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L222-L223). 

These two cases do not agree with each other. But more often we view the later one as an issue since it make an exception that `realm` can be `null` and has been causing issues in handling it. Therefore I overall feel it's OK to fail earlier here. Letting it passing through is not actually useful other than for argument sake.

The practical bottom line is that it is not possible to run-as API keys. So not a real issue anyway.

Reply:
Thanks for raising this topic. FWIW My reasoning was that both users (authenticating and run-as) should have access to a particular workflow.  Allowing the authenticating user to invoke a workflow should be prevented early on, even if run-as user can access it.

Additionally, what we could potentially do is differentiate between which role was restricted, but this felt unnecessary at the moment.

Reply:
> Allowing the authenticating user to invoke a workflow should be prevented early on, even if run-as user can access it.

My view is the opposite. When
1. The API is a workflow API 
2. The authenticating user has the privilege to run-as the effective user
3. The effective user has a workflow role that allows it to call the workflow API

In this situation, the authorization should succeed. I believe that's what naturally would happen with the proposed change. Because Item 2 implies the authenticating user must have access to the workflow as well, either because the authenticating user has a workflow role or has a role _without_ workflow restriction. 

My original comment was about the failure case. That is, Item 1 & 2 remain unchanged, but in Item 3, the effective user does *not* have access to the workflow API. In this case, we could let the run-as (Item 2) to be successful (it generates an audit log entry `runAsGranted`), then fail due to Item 3. Or we can fail before even check Item 2 (and skip the `runAsGranted` audit entry), which is the behaviour of the proposed change. Both options are valid to some extend. But going with the later option seems to give us more flexibility in future (thinking moving rejection to the REST layer).